skyportal/handlers/api/analysis.py

```diff
@@ -0,0 +1,423 @@
+                    - type: object
+                      properties:
+                        data:
+                          type: object
+                          properties:
+                            id:
+                              type: integer
+                              description: New AnalysisService ID
+        """
+        data = self.get_json()
+
+        if not data.get('url', None):
+            return self.error('`url` is required to add an Analysis Service.')
+        else:
+            if not valid_url(data.get('url', None)):
+                return self.error(
+                    'a valid `url` is required to add an Analysis Service.'
+                )
+
+        authentication_type = data.get('authentication_type', None)
+        if not authentication_type:
+            return self.error(
+                '`authentication_type` is required to add an Analysis Service.'
+            )
+
+        if authentication_type not in AUTHENTICATION_TYPES:
+            return self.error(
+                f'`authentication_type` must be one of: {", ".join([t for t in AUTHENTICATION_TYPES])}.'
+            )
+        else:
+            if authentication_type != 'none':
+                _authinfo = data.get('_authinfo', None)
+                if not _authinfo:
+                    return self.error(
+                        '`_authinfo` is required to add an Analysis Service '
+                        ' when authentication_type is not "none".'
+                    )
+                try:
+                    _authinfo = json.loads(_authinfo)
+                except json.JSONDecodeError:
+                    return self.error(
+                        '`_authinfo` must be parseable to a valid JSON object.'
+                    )
+                if authentication_type not in _authinfo:
+                    return self.error(
+                        f'`_authinfo` must contain a key for "{authentication_type}".'
+                    )
+
+        group_ids = data.pop('group_ids', None)
+        if not group_ids:
+            groups = self.current_user.accessible_groups
+        else:
+            try:
+                groups = Group.get_if_accessible_by(
+                    group_ids, self.current_user, raise_if_none=True
+                )
+            except AccessError:
+                return self.error('Could not find any accessible groups.', status=403)
+
+        schema = AnalysisService.__schema__()
+        try:
+            analysis_service = schema.load(data)
+        except ValidationError as e:
+            return self.error(
+                "Invalid/missing parameters: " f"{e.normalized_messages()}"
+            )
+
+        DBSession().add(analysis_service)
+        analysis_service.groups = groups
+
+        try:
+            self.verify_and_commit()
+        except IntegrityError as e:
+            return self.error(
+                f'Analysis Service with that name already exists: {str(e)}'
+            )
```

Comment:
Is this the only integrity error that can come up? Wouldn't it be safer to just propagate the error message from the DB?

Reply:
done.